### PR TITLE
SI-8442 Ignore stub annotation symbols in `AnnotationInfo#matches`

### PR DESCRIPTION
**Pull request by retronym**
And update the java `ClassFileParser` to create distinguished
`StubClassSymbol`s, rather that a regular `ClassSymbol`s, when
encountering a deficient classpath. This brings it into line
with `Unpickler`, which has done as much since a55788e275f.

This stops the enclosed test case from crashing when determining
if the absent symbol, `A_1`, is a subclass of `@deprecated`.

This is ostensibly fixes a regression, although it only worked in
`2.10.[0-3]` by a fluke: the class file parser's promiscious
exception handling caught and recovered from the NPE introduced
in SI-7439!

```
% javac -d /tmp test/files/run/t8442/{A,B}_1.java &&  qbin/scalac -classpath /tmp -d /tmp test/files/run/t8442/C_2.scala && (rm /tmp/A_1.class; true) && scalac-hash v2.10.0 -classpath /tmp -d /tmp test/files/run/t8442/C_2.scala
warning: Class A_1 not found - continuing with a stub.
warning: Caught: java.lang.NullPointerException while parsing annotations in /tmp/B_1.class
two warnings found
```

Review by @gkossakowski

### CODE REVIEW

File: test/files/run/t8442/Test.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import scala.tools.partest._
+import java.io.File
+
+object Test extends StoreReporterDirectTest {
+  def code = ???
+
+  def compileCode(code: String) = {
+    val classpath = List(sys.props("partest.lib"), testOutput.path) mkString sys.props("path.separator")
+    compileString(newCompiler("-cp", classpath, "-d", testOutput.path))(code)
```
Comment (gkossakowski) [inline test/files/run/t8442/Test.scala:9]: Are those two lines needed to support separate compilation?

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala [hunk 2]
```diff
@@ -436,9 +436,13 @@ abstract class ClassfileParser {
...
       // that are not in their correct place (see bug for details)
-      if (!settings.isScaladoc)
-        warning(s"Class $name not found - continuing with a stub.")
-      return NoSymbol.newClass(name.toTypeName)
+
+      // TODO More consistency with use of stub symbols in `Unpickler`
+      //   - better owner than `NoSymbol`
+      //   - remove eager warning
+      val msg = s"Class $name not found - continuing with a stub."
+      if (!settings.isScaladoc) warning(msg)
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:444]: Why do we need the special case for scaladoc?

Comment (retronym) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:444]: That's existing code, supported by the comment above.

File: test/files/run/t8442/Test.scala (hunk 1 above)
Comment (retronym) [inline test/files/run/t8442/Test.scala:9]: Yep, we need to see the results of `javac`.

File: src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala (hunk 2 above)
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala:444]: Yes, I saw it right after sending the comment (and then my internet connection died).

File: test/files/run/t8442/Test.scala (hunk 1 above)
Comment (gkossakowski) [inline test/files/run/t8442/Test.scala:9]: It would be great to add a comment documenting that.

**Comment by gkossakowski** [comment]
LGTM

**Pull request merged**